### ISSUE TITLE
Add option to prevent `writers.gdal` to fail if there are no points to write

### ISSUE THREAD

**Issue by zdila** [PDAL/PDAL#3977]
**Describe the bug**

Add option to prevent `writers.gdal` to fail if there are no points to write.

<details>
  
```
[
  "work_alt0/36_Slanec_21_528487_5390439_c_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_528635_5390308_b_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_529139_5387186_e_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_529754_5393179_d_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_529581_5389254_b_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_530090_5395374_c_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_530151_5384680_e_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_530480_5386927_b_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_530761_5387896_d_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_531239_5388378_b_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_531325_5393392_c_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_531741_5385591_d_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_532020_5389177_b_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_532314_5390338_c_c_jtsk03_bpv.laz",
  "work_alt0/36_Slanec_21_533816_5390562_b_c_jtsk03_bpv.laz",
  {
    "type": "writers.gdal",
    "resolution": 0.5,
    "filename": "work_alt0/binary_3.tif",
    "gdalopts": "COMPRESS=DEFLATE,PREDICTOR=2,ZLEVEL=5",
    "output_type": "count",
    "data_type": "uint16",
    "where": "(Classification == 3)"
  },
  {
    "type": "writers.gdal",
    "resolution": 0.5,
    "filename": "work_alt0/binary_4.tif",
    "gdalopts": "COMPRESS=DEFLATE,PREDICTOR=2,ZLEVEL=5",
    "output_type": "count",
    "data_type": "uint16",
    "where": "(Classification == 4)"
  },
  {
    "type": "writers.gdal",
    "resolution": 0.5,
    "filename": "work_alt0/binary_5.tif",
    "gdalopts": "COMPRESS=DEFLATE,PREDICTOR=2,ZLEVEL=5",
    "output_type": "count",
    "data_type": "uint16",
    "where": "(Classification == 5)"
  },
  {
    "type": "writers.gdal",
    "resolution": 0.5,
    "filename": "work_alt0/binary_6.tif",
    "gdalopts": "COMPRESS=DEFLATE,PREDICTOR=2,ZLEVEL=5",
    "output_type": "count",
    "data_type": "uint16",
    "where": "(Classification == 6)"
  },
  {
    "type": "writers.gdal",
    "resolution": 0.5,
    "filename": "work_alt0/binary_9.tif",
    "gdalopts": "COMPRESS=DEFLATE,PREDICTOR=2,ZLEVEL=5",
    "output_type": "count",
    "data_type": "uint16",
    "where": "(Classification == 9)"
  }
]
```

</details>

**Expected behavior**

There should be additional option for `writers.gdal` that would prevent `pdal` to error with `PDAL: Unable to write GDAL data with no points for output.` if there is no point to write. In that case it would just not write the file at all. My use-case is that I have multiple `writers.gdal` in my pipeline where each has `where` to filter points only of specific Classification and there are many "tiles" not containing points of all classifications.

My workaround is to replace `throw` with `return` in the source code.

**System/installation information:**
Please provide information on your PDAL version (`pdal --version`) and system (e.g., `uname -a` or `ver`).

```
$ uname -a
```

`Linux bono 6.1.0-3-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.8-1 (2023-01-29) x86_64 GNU/Linux`

```
$ pdal --version
```

`pdal 2.3.0 (git-version: Release)`

`pdal 2.6.0 (git-version: 3d3db4)`

**Comment by hobu** [comment]
> There should be additional option for writers.gdal that would prevent pdal to error with PDAL: Unable to write GDAL data with no points for output. 

We do not do this for any writer type, so doing it for just `writers.gdal` would be different behavior than all of the rest of them. We were previously complaining with a specific and user unfriendly error that meant there were no points.

There's two kinds of empty here:
1) We know the bounds of the raster and we simply are not writing any points (your scenario)
2) We actually have zero points and do not have any bounding information about the raster to even construct a valid container.

Your patch for the first one, and an option like `writers.gdal.allow_empty=True` to enable it is probably reasonable. 

The second scenario is much harder, however. In streaming mode, I do not think an option like `writers.gdal.allow_empty=True` could work, because we need to either know the dimensions of the raster we are writing beforehand (like standard mode) or accumulate it as we write points. Otherwise when we ask GDAL to make an empty raster, it is going to blow up. We don't have the information that we are empty until the end of the process

**Comment by zdila** [comment]
My hack ([`throw`](https://github.com/PDAL/PDAL/blob/3d3db44b79a53a7f134bcdd90f8654fa537c8516/io/GDALWriter.cpp#L286) replaced by `return`) causes creating no file at all. This is sufficient for my use-case where I don't need empty (nodata-only) tiff.

**Comment by hobu** [comment]
> My hack ([throw](https://github.com/PDAL/PDAL/blob/3d3db44b79a53a7f134bcdd90f8654fa537c8516/io/GDALWriter.cpp#L286) replaced by return) causes creating no file at all. This is sufficient for my use-case where I don't need empty (nodata-only) tiff.

yes, your scenario is the first as I've presented above. I hope other developers have opinions about the second.

**Comment by hobu** [comment]
Part 1 addressed in #3984 3984

**Comment by hobu** [comment]
#3984 implements this for now. reopen if we still have issues.
